### PR TITLE
Fix replication

### PR DESCRIPTION
**Pull request by haadcode**
This PR fixes replication in OrbitDB.

Long ago we made the changes in ipfs-log to support parallel fetching of entries but those updates never made it into OrbitDB side. This PR adds those changes thus adding full replication concurrency support to OrbitDB. This in turn means that databases are replicated faster than before in addition to loading faster than before.

This has a major impact on the perceived (and actual) speed of OrbitDB 🎉

This PR supersedes #929 as it was easier to create a new branch and PR given the commit history of this change.

This PR depends on orbitdb/orbit-db-store#117, orbitdb/ipfs-log#340 and https://github.com/orbitdb/orbit-db-counterstore/pull/44.

### CODE REVIEW

**Comment by haadcode** [comment]
This PR is now ready for review!

The PRs that this PR depends on should be reviewed and merged first in the following order:
- orbitdb/ipfs-log#340
- orbitdb/orbit-db-counterstore#44
- orbitdb/orbit-db-store#117

Happy to clarify on any of the changes and dependency PRs!

File: src/OrbitDB.js [hunk 1]
```diff
@@ -32,6 +32,8 @@ const databaseTypes = {
   keyvalue: KeyValueStore
 }
 
+const defaultTimeout = 30000 // 30 seconds
```
[the remark this answers is not in the document]
Comment (tabcat) [reply src/OrbitDB.js:35]: orbit-db-io@1.0.2 passes the option correctly now but it appears like the js-ipfs-http-api doesnt have a working timeout option for the block api https://github.com/orbitdb/orbit-db-io/pull/29

File: test/create-open.test.js [hunk 2]
```diff
@@ -352,7 +352,7 @@ Object.keys(testAPIs).forEach(API => {
       })
 
       it('doesn\'t open a database if we don\'t have it locally', async () => {
-        const db = await orbitdb.open('abc', { create: true, type: 'feed', overwrite: true })
+        const db = await orbitdb.open('abcabc', { create: true, type: 'feed', overwrite: true })
```
[the remark this answers is not in the document]
Comment (haadcode) [reply test/create-open.test.js:355]: Correct, to make sure we get an empty db for the test.

File: src/OrbitDB.js [hunk 3]
```diff
@@ -104,7 +106,7 @@ class OrbitDB {
     }
 
     if (!options.keystore) {
-      const keystorePath = path.join(options.directory, id, '/keystore')
+      const keystorePath = path.join(options.directory, options.id || id, '/keystore')
```
[the remark this answers is not in the document]
Comment (haadcode) [reply src/OrbitDB.js:109]: Will fix 👍

File: src/OrbitDB.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (haadcode) [reply src/OrbitDB.js:35]: 👍

**Review by tabcat — APPROVED** [review]
:tada:

**Comment by haadcode** [comment]
The dependency PRs were merged:
- https://github.com/orbitdb/orbit-db-store/pull/117
- https://github.com/orbitdb/ipfs-log/pull/340
- https://github.com/orbitdb/orbit-db-counterstore/pull/44

Review feedback addressed and rest are passing 👌 Going to merge.

**Comment by haadcode** [comment]
Published to npm as `0.28.0`

**Comment by holmesworcester** [comment]
@haadcode we're seeing a regression (or some unreliability in replication) here: https://github.com/orbitdb/orbit-db-store/issues/122

Is it possible this is the result of the refactor? Any thoughts on what might fix it?

**Pull request merged**